### ISSUE TITLE
Does not allow use of general options in the repl

### ISSUE THREAD

**Issue by KSchopmeyer** [click-contrib/click-repl#103]
If a command consists of general options (i.e. options before the command name), click_repl version 2.0  supported modifying them as part of a repl entered command.

Thus, if the app included a general option --verbose and the repl command looks like this "--verbose <my command> <possible arguments> <possible options>" click-repl accepted the --verbose general cmd line option.click-repl V2  allowed the use
both the options before the command and of tab-completion for these options and their values if the app code implements the tab-completion.

Right now,  using the master branch entering any of these general options in the interactive mode ends up in a series of exceptions as shown below with the option to continue and eventually the command can be entered.

Example follows:

pywbemcli -n mock1
Enter 'help repl' for help, <CTRL-D> or ':q' to exit pywbemcli or <CTRL-r> to search history, 

Unhandled exception in event loop:
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/buffer.py", line 1933, in new_coroutine
    await coroutine(*a, **kw)
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/buffer.py", line 1757, in async_completer
    async for completion in async_generator:
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/completion/base.py", line 323, in get_completions_async
    async for completion in completer.get_completions_async(
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/completion/base.py", line 199, in get_completions_async
    for item in self.get_completions(document, complete_event):
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click_repl/_completer.py", line 257, in get_completions
    self.parsed_ctx = _resolve_context(args, self.ctx)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click_repl/utils.py", line 46, in _resolve_context
    name, cmd, args = command.resolve_command(ctx, args)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click/core.py", line 1715, in resolve_command
    ctx.fail(_("No such command {name!r}.").format(name=original_cmd_name))
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click/core.py", line 679, in fail
    raise UsageError(message, self)

Exception No such command '-n'.
Press ENTER to continue...

Unhandled exception in event loop:
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/buffer.py", line 1933, in new_coroutine
    await coroutine(*a, **kw)
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/buffer.py", line 1757, in async_completer
    async for completion in async_generator:
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/completion/base.py", line 323, in get_completions_async
    async for completion in completer.get_completions_async(
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/completion/base.py", line 199, in get_completions_async
    for item in self.get_completions(document, complete_event):
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click_repl/_completer.py", line 257, in get_completions
    self.parsed_ctx = _resolve_context(args, self.ctx)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click_repl/utils.py", line 46, in _resolve_context
    name, cmd, args = command.resolve_command(ctx, args)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click/core.py", line 1715, in resolve_command
    ctx.fail(_("No such command {name!r}.").format(name=original_cmd_name))
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click/core.py", line 679, in fail
    raise UsageError(message, self)

Exception No such command '-n'.
Press ENTER to continue...

Unhandled exception in event loop:
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/buffer.py", line 1933, in new_coroutine
    await coroutine(*a, **kw)
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/buffer.py", line 1757, in async_completer
    async for completion in async_generator:
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/completion/base.py", line 323, in get_completions_async
    async for completion in completer.get_completions_async(
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/prompt_toolkit/completion/base.py", line 199, in get_completions_async
    for item in self.get_completions(document, complete_event):
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click_repl/_completer.py", line 257, in get_completions
    self.parsed_ctx = _resolve_context(args, self.ctx)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click_repl/utils.py", line 46, in _resolve_context
    name, cmd, args = command.resolve_command(ctx, args)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click/core.py", line 1715, in resolve_command
    ctx.fail(_("No such command {name!r}.").format(name=original_cmd_name))
  File "/home/kschopmeyer/virtualenvs/pycli311/lib/python3.11/site-packages/click/core.py", line 679, in fail
    raise UsageError(message, self)

Exception No such command '-n'.
Press ENTER to continue...
pywbemcli> -n mock3ns class enumerate
Usage: pywbemcli [GENERAL-OPTIONS] COMMAND [ARGS] [COMMAND-OPTIONS]

Error: No such command '-n'.

**Comment by KSchopmeyer** [comment]
Since this worked in click-repl version 2.0, this is a serious backward incompatibility and actually makes this new release unusable to at least one of our projects.

**Comment by auvipy** [comment]
@GhostOps77 I guess you might know about the regression?

**Comment by KSchopmeyer** [comment]
I do know of the regression and it means that we are going to have to either stay with v2 or put the support of the general options back in ourselves.  And with the long times to release things like click_repl, it will probably meaning creating a fork or using a vendorzed version of click_repl.  I test this summer with our test suite failed over a hundred of the test largely because we use the general options in the repl mode (A very simple example the --verbose option).

**Comment by andy-maier** [comment]
I was digging into this a bit, and found that PR #97 was the change that caused global options to no longer work.

If I change the code of what is now (master branch) in `src/_repl.py` line 157 from:

```
                    # The group command will dispatch based on args.
                    old_protected_args = _get_protected_args(group_ctx)
                    try:
                        _set_protected_args(group_ctx, args)
                        group.invoke(group_ctx)
                    finally:
                        _set_protected_args(group_ctx, old_protected_args)
```

to what it was before PR #97:

```
                    # default_map passes the top-level params to the new group to
                    # support top-level required params that would reject the
                    # invocation if missing.
                    with group.make_context(
                        None, args, parent=group_ctx, default_map=old_ctx.params
                    ) as ctx:
                        group.invoke(ctx)
                        ctx.exit()
```

then using global options in repl mode works again, e.g. with the zhmccli project:
* The shell command `zhmc -o json` sets the output format to JSON and enters repl mode.
* The repl command `cpc list` inherits that output format from the command line.
* The repl command `-o table cpc list` overwrites the inherited output format.

```
$ zhmc -o json
Enter 'help' for help, <CTRL-D> or ':q' to exit.

zhmc> cpc list
[{ ... (output not shown) ... }]

zhmc> -o table cpc list
+--------+------------------+---------------+--------------+----------------+-----------------+-------------------------+---------------------+
| name   | status           | dpm-enabled   | se-version   | machine-type   | machine-model   | machine-serial-number   | description         |
|--------+------------------+---------------+--------------+----------------+-----------------+-------------------------+---------------------|
. . . (output not shown)
+--------+------------------+---------------+--------------+----------------+-----------------+-------------------------+---------------------+

zhmc>
```
As you can see, the output format is as expected for each of the two commands.

I also tried the code shown by @flacoste in issue #92, in the version where `db` is an option, with two additional fixes:
* Added `@classmethod` to `MyDB.get_entry()`
* Used some other defined exception instead of `NotFound` in `entry()`

That version worked with both versions of the `src/_repl.py` code shown above (i.e. master and before PR #97) as @flacoste would have expected:

```
$ ./flacoste_shell.py --db db entry 1
ENTRY 1
```

```
$ ./flacoste_shell.py --db db
> entry 1
ENTRY 1
>
```

When using the version of the code where `db` is a positional argument, it requires the fix in PR #97 to work. However, that fix causes global options not to be useable in repl mode.

My question is: Where do we go from here, what are the use cases we want to support?

@flacoste @GhostOps77 @auvipy - your comments are welcome

**Comment by auvipy** [comment]
@andy-maier can you please check the main branch again? and tell me if any pr need to be reverted? or if you can contribute of any exisiting regression?

**Comment by andy-maier** [comment]
Hi. I'll check it out soon. Thanks!

**Comment by auvipy** [comment]
@andy-maier would be grateful if you can share your feedback.... as I have to release a new version or two in this month.... this package is used by celery as well...

**Comment by andy-maier** [comment]
@auvipy Hello Asif. I have tested the master branch as of commit dc09e1b4 (with click 8.5.0).

The result was that the exception we got with click-repl 0.3.0 when starting the command input with a global option no longer happens, so that is definitely an improvement.

The behavior for looking up the values of global options that have completion support differs a bit between 0.2.0 and the master branch:
* With 0.2.0, the option was treated as if it had no value, and the global options and commands were offered as choices.
* With the master branch, trying to complete the option value resulted in offering nothing (for comparison, when doing the same in the terminal (i.e. without interactive repl mode), the option values were properly looked up and offered as choices).

Since completion of option values does not work in either 0.2.0 or master, this may well be an issue on our end that we need to investigate.

Bottom line: Since the exception no longer happens, I think the click-repl code is good enough to be released, from my view.

We'll discuss this on monday in our dev team, and in case there are any new thoughts on ths, I'll let you know. So please hold back with the release until tuesday.

**Comment by auvipy** [comment]
ok sure. thank you

**Comment by andy-maier** [comment]
@auvipy Hello Asif.
We discussed this in our dev team and we think you should go ahead and make the release.

**Comment by auvipy** [comment]
OK thanks a lot!

**Comment by andy-maier** [comment]
We think this issue is sufficiently resolved in 0.4.0.

**Comment by KSchopmeyer** [comment]
closed effective release 0.4.0
